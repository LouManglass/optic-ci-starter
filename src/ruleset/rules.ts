import {expect} from "chai";
import {ApiChangeDsl} from "@useoptic/api-checks";

export const rules = {

  operationIdRules: ({operations}: ApiChangeDsl) => {
    operations.changed.must(
      'have consistent operation IDs',
      (current, next) => {
        expect(current.operationId).to.equal(next.operationId);
      }
    );
  },

  propertyRules: ({bodyProperties}: ApiChangeDsl) => {

    function isValidKeyName(key: string) {
      return true
    }

    bodyProperties.added.must('following naming convention', (property) => {
      if (!isValidKeyName(property.key)) expect.fail(`added property ${property.key} must be your_case`)
    })

    bodyProperties.changed.must('request properties should never become required', (before, after, context) => {
      if ('inRequest' in context) {

        if (!before.required && after.required) {
          expect.fail(`${before.key} can not become required`)
        }

      }
    })

    bodyProperties.changed.must('response properties should never become optional', (before, after, context) => {
      if ('inResponse' in context) {

        if (before.required && !after.required) {
          expect.fail(`${before.key} can not become optional`)
        }

      }
    })

  }
}
