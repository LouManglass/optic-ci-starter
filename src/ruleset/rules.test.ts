import {ApiChangeDsl, ApiCheckDsl, ApiCheckDslContext, createTestDslFixture} from "@useoptic/api-checks";
import {rules} from "./rules";

export function createFixture() {
  return createTestDslFixture<ApiChangeDsl, ApiCheckDslContext>((input) =>
    new ApiChangeDsl(input.nextFacts, input.changelog, input.currentJsonLike, input.nextJsonLike, input.context))
}

describe("operation ids", () => {
  const baseForOperationIdTests = {
    openapi: '3.0.1',
    paths: {
      '/example': {
        operationId: "before",
        get: {
          responses: {},
        },
      },
    },
    info: { version: '0.0.0', title: 'Empty' },
  };

  it("fails if changed", async () => {
    const {compare} = createFixture()
    const result = await compare(baseForOperationIdTests)
      .to((spec) => {
        spec.paths!["/example"]!.get!.operationId = "changed-from-before";
        return spec;
      })
      .withRule(rules.operationIdRules, {});

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });
  it("passes if no operation id in from spec", async () => {
    const {compare} = createFixture()

    const noOperationId = {
      openapi: '3.0.1',
      paths: {
        '/example': {
          get: {
            responses: {},
          },
        },
      },
      info: { version: '0.0.0', title: 'Empty' },
    };

    const result = await compare(noOperationId)
      .to((spec) => {
        spec.paths!["/example"]!.get!.operationId = "changed-from-empty";
        return spec;
      })
      .withRule(rules.operationIdRules, {});

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });
})
