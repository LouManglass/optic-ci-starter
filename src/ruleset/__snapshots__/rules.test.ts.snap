// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation ids fails if empty string 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
        "operationId": "before",
      },
    },
  },
  "changelog": Array [
    Object {
      "changeType": "changed",
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "changed-from-before",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "changed-from-before",
          "responses": Object {},
        },
        "operationId": "before",
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "method": "get",
            "operationId": "changed-from-before",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected undefined to equal 'changed-from-before'",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;

exports[`operation ids passes if no operation id in from spec 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changeType": "changed",
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "changed-from-empty",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "changed-from-empty",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "method": "get",
            "operationId": "changed-from-empty",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected undefined to equal 'changed-from-empty'",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;
